File: android/app/src/main/java/com/mattermost/rnbeta/RNPasteableActionCallback.java
package com.mattermost.rnbeta;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.net.Uri;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuItem;

public class RNPasteableActionCallback implements ActionMode.Callback {

    private RNPasteableEditText mEditText;

    RNPasteableActionCallback(RNPasteableEditText editText) {
        mEditText = editText;
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        if (item.getItemId() == android.R.id.paste && this.shouldCustomHandlePaste()) {
            mEditText.getOnPasteListener().onPaste();
            mode.finish();
        } else {
            mEditText.onTextContextMenuItem(item.getItemId());
        }

        return true;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {

    }

    private boolean shouldCustomHandlePaste() {
        ClipboardManager clipboardManager = (ClipboardManager) mEditText.getContext().getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clipData = clipboardManager.getPrimaryClip();

        ClipData.Item item = clipData.getItemAt(0);
        if (item == null) {
            return false;
        }

        Uri itemUri = item.getUri();
        return itemUri != null;
    }
}

File: android/app/src/main/java/com/mattermost/rnbeta/RNPasteableActionCallback.kt
package com.mattermost.rnbeta

import android.R
import android.content.ClipboardManager
import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.view.ActionMode
import android.view.Menu
import android.view.MenuItem
import com.facebook.react.bridge.Arguments

class RNPasteableActionCallback internal constructor(private val mEditText: RNPasteableEditText) : ActionMode.Callback {
    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        val config: Bundle = com.mattermost.rnbeta.MainApplication.instance.getManagedConfig()
        val result = Arguments.fromBundle(config)
        val copyPasteProtection = result.getString("copyAndPasteProtection")
        if (copyPasteProtection == "true") {
            disableMenus(menu)
        }
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean = false

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        val uri = uriInClipboard
        if (item.itemId == R.id.paste && uri != null) {
            mEditText.onPasteListener?.onPaste(uri)
            mode.finish()
        } else {
            mEditText.onTextContextMenuItem(item.itemId)
        }
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) {}
    private fun disableMenus(menu: Menu) {
        for (i in 0 until menu.size()) {
            val item = menu.getItem(i)
            val id = item.itemId
            val shouldDisableMenu = id == R.id.paste || id == R.id.copy || id == R.id.cut
            item.isEnabled = !shouldDisableMenu
        }
    }

    private val uriInClipboard: Uri?
        get() {
            val clipboardManager = mEditText.context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clipData = clipboardManager.primaryClip ?: return null
            val item = clipData.getItemAt(0) ?: return null
            val text = item.text.toString()
            return if (text.length > 0) {
                null
            } else item.uri
        }

}
